File: src/utilities/welcome_image.py
from PIL import Image, ImageDraw, ImageFont
import requests
from io import BytesIO

#* Función para cargar la imagen base
def cargar_imagen_base(base_image_path):
    try:
        return Image.open(base_image_path)
    except Exception as e:
        print(f"Error al cargar la imagen base: {e}")
        return None


#* Función para obtener y redimensionar el avatar del usuario
def obtener_avatar(avatar_url, size=(200, 200)):
    try:
        response = requests.get(avatar_url)
        avatar_image = Image.open(BytesIO(response.content)).resize(size)
        return avatar_image
    except Exception as e:
        print(f"Error al cargar el avatar: {e}")
        return None


#* Función para hacer el avatar circular con fondo transparente
def make_round_avatar(avatar_image, size=(200, 200)):
    avatar_mask = Image.new("L", size, 0)
    draw_mask = ImageDraw.Draw(avatar_mask)
    draw_mask.ellipse((0, 0, *size), fill=255)
    avatar_image = avatar_image.convert("RGBA")
    avatar_image.putalpha(avatar_mask)
    return avatar_image


#* Función para añadir un borde circular al avatar
def add_border_avatar(avatar_image, border_size=6, border_color=(255, 214, 91)):
    avatar_with_border = Image.new("RGBA", (avatar_image.width + border_size * 2, avatar_image.height + border_size * 2), (0, 0, 0, 0))
    draw = ImageDraw.Draw(avatar_with_border)
    draw.ellipse([0, 0, avatar_with_border.width, avatar_with_border.height], fill=border_color)
    avatar_with_border.paste(avatar_image, (border_size, border_size), avatar_image)
    return avatar_with_border


#* Función para cargar la fuente personalizada
def load_font(font_path, font_size):
    try:
        return ImageFont.truetype(font_path, font_size)
    except Exception as e:
        print(f"Error al cargar la fuente: {e}")
        return None


#* Función para escribir el texto de bienvenida
def write_welcome_text(draw, nombre_usuario, font, base_image_height, x_start=250):
    text_bienvenido = "¡Bienvenid@, "
    text_exclamacion = "!"
    
    # Calcular la posición vertical centrada para el texto
    text_height = font.getbbox(text_bienvenido)[1]
    y_centered = (base_image_height - text_height) // 2 - 30
    y_start = y_centered

    # Escribir los textos
    draw.text((x_start, y_start), text_bienvenido, (255, 214, 91), font=font)
    text_bienvenido_width = font.getbbox(text_bienvenido)[2]

    # Escribir el nombre del usuario
    x_new = x_start + text_bienvenido_width
    draw.text((x_new, y_start), nombre_usuario, (224, 168, 0), font=font)
    text_usuario_width = font.getbbox(nombre_usuario)[2]

    # Escribir el signo de exclamación
    x_new += text_usuario_width
    draw.text((x_new, y_start), text_exclamacion, (255, 214, 91), font=font)


#* Función principal para generar la imagen de bienvenida
def generate_welcome_image(nombre_usuario, avatar_url, base_image_path):
    # Cargar imagen base
    base_image = cargar_imagen_base(base_image_path)
    if base_image is None:
        return None

    # Obtener avatar y procesarlo
    avatar_image = obtener_avatar(avatar_url)
    if avatar_image is None:
        return None
    
    avatar_image = make_round_avatar(avatar_image)
    avatar_with_border = add_border_avatar(avatar_image)

    # Cargar fuente personalizada
    font = load_font("assets/fonts/Comfortaa-Bold.ttf", 60)
    if font is None:
        return None

    # Crear objeto de dibujo y escribir texto
    draw = ImageDraw.Draw(base_image)
    write_welcome_text(draw, nombre_usuario, font, base_image.size[1])

    # Pegar el avatar en la imagen base
    avatar_position_y = (base_image.size[1] - avatar_with_border.size[1]) // 2
    base_image.paste(avatar_with_border, (25, avatar_position_y), avatar_with_border)

    # Guardar la imagen final en un buffer y retornarla
    image_binary = BytesIO()
    try:
        base_image.save(image_binary, 'PNG')
        image_binary.seek(0)
    except Exception as e:
        print(f"Error al guardar la imagen: {e}")
        return None

    return image_binary

File: src/utilities/welcome_image.js
import { createCanvas, loadImage, registerFont } from 'canvas';
import fetch from 'node-fetch';
import fs from 'fs';
import path from 'path';

// MARK: Cargar imagen base
/**
 * Función para cargar la imagen base
 * @param {string} baseImagePath - Ruta de la imagen base
 * @returns {Promise<Image|null>} - Imagen cargada o null si hay error
 */
async function cargarImagenBase(baseImagePath) {
    try {
        return await loadImage(baseImagePath);
    } catch (error) {
        console.error(`Error al cargar la imagen base: ${error}`);
        return null;
    }
}

// MARK: Obtener avatar del usuario
/**
 * Función para obtener y redimensionar el avatar del usuario
 * @param {string} avatarUrl - URL del avatar del usuario
 * @param {Array<number>} size - Tamaño del avatar [width, height]
 * @returns {Promise<Image|null>} - Avatar redimensionado o null si hay error
 */
async function obtenerAvatar(avatarUrl, size = [200, 200]) {
    try {
        const response = await fetch(avatarUrl);
        const arrayBuffer = await response.arrayBuffer();
        const buffer = Buffer.from(arrayBuffer);
        const avatarImage = await loadImage(buffer);

        // Crear canvas para redimensionar
        const canvas = createCanvas(size[0], size[1]);
        const ctx = canvas.getContext('2d');
        ctx.drawImage(avatarImage, 0, 0, size[0], size[1]);

        return await loadImage(canvas.toBuffer());
    } catch (error) {
        console.error(`Error al cargar el avatar: ${error}`);
        return null;
    }
}

// MARK: Avatar circular con borde
/**
 * Función para hacer el avatar circular con borde
 * @param {Image} avatarImage - Imagen del avatar
 * @param {Array<number>} size - Tamaño del avatar [width, height]
 * @param {number} borderSize - Tamaño del borde
 * @param {string} borderColor - Color del borde
 * @returns {Canvas} - Canvas con el avatar circular
 */
function makeRoundAvatarWithBorder(avatarImage, size = [200, 200], borderSize = 6, borderColor = '#00ace2') {
    const totalSize = [size[0] + borderSize * 2, size[1] + borderSize * 2];
    const canvas = createCanvas(totalSize[0], totalSize[1]);
    const ctx = canvas.getContext('2d');

    // Crear círculo con borde
    ctx.beginPath();
    ctx.arc(totalSize[0] / 2, totalSize[1] / 2, totalSize[0] / 2, 0, Math.PI * 2);
    ctx.fillStyle = borderColor;
    ctx.fill();

    // Crear máscara circular para el avatar
    ctx.save();
    ctx.beginPath();
    ctx.arc(totalSize[0] / 2, totalSize[1] / 2, size[0] / 2, 0, Math.PI * 2);
    ctx.clip();

    // Dibujar el avatar dentro del círculo
    ctx.drawImage(avatarImage, borderSize, borderSize, size[0], size[1]);
    ctx.restore();

    return canvas;
}

// MARK: Fuente
/**
 * Función para cargar la fuente personalizada
 * @param {string} fontPath - Ruta de la fuente
 * @param {string} fontFamily - Nombre de la familia de fuentes
 */
function loadFont(fontPath, fontFamily) {
    try {
        if (fs.existsSync(fontPath)) {
            registerFont(fontPath, { family: fontFamily });
            return true;
        } else {
            console.warn(`Fuente no encontrada: ${fontPath}`);
            return false;
        }
    } catch (error) {
        console.error(`Error al cargar la fuente: ${error}`);
        return false;
    }
}

// MARK: Texto de bienvenida
/**
 * Función para escribir el texto de bienvenida
 * @param {CanvasRenderingContext2D} ctx - Contexto del canvas
 * @param {string} nombreUsuario - Nombre del usuario
 * @param {string} font - Especificación de la fuente
 * @param {number} baseImageHeight - Altura de la imagen base
 * @param {number} xStart - Posición X inicial
 */
function writeWelcomeText(ctx, nombreUsuario, font, baseImageHeight, xStart = 250) {
    const textBienvenido = '¡Bienvenid@, ';
    const textExclamacion = '!';

    // Configurar fuente
    ctx.font = font;

    // Calcular la posición vertical centrada para el texto
    const textMetrics = ctx.measureText(textBienvenido);
    const textHeight = textMetrics.actualBoundingBoxAscent + textMetrics.actualBoundingBoxDescent;
    const yCentered = (baseImageHeight / 2) + (textMetrics.actualBoundingBoxAscent / 2);
    const yStart = yCentered;

    // Escribir "¡Bienvenid@, "
    ctx.fillStyle = '#00ace2'; // Color dorado
    ctx.fillText(textBienvenido, xStart, yStart);
    const textBienvenidoWidth = ctx.measureText(textBienvenido).width;

    // Escribir el nombre del usuario
    const xNew = xStart + textBienvenidoWidth;
    ctx.fillStyle = '#005670ff'; // Color dorado más oscuro
    ctx.fillText(nombreUsuario, xNew, yStart);
    const textUsuarioWidth = ctx.measureText(nombreUsuario).width;

    // Escribir el signo de exclamación
    const xFinal = xNew + textUsuarioWidth;
    ctx.fillStyle = '#00ace2'; // Color dorado
    ctx.fillText(textExclamacion, xFinal, yStart);
}

// MARK: Generar imagen de bienvenida
/**
 * Función principal para generar la imagen de bienvenida
 * @param {string} nombreUsuario - Nombre del usuario
 * @param {string} avatarUrl - URL del avatar del usuario
 * @param {string} baseImagePath - Ruta de la imagen base
 * @returns {Promise<Buffer|null>} - Buffer de la imagen generada o null si hay error
 */
export async function generateWelcomeImage(nombreUsuario, avatarUrl, baseImagePath) {
    try {
        // Cargar imagen base
        const baseImage = await cargarImagenBase(baseImagePath);
        if (!baseImage) {
            return null;
        }

        // Obtener avatar y procesarlo
        const avatarImage = await obtenerAvatar(avatarUrl);
        if (!avatarImage) {
            return null;
        }

        // Crear canvas principal con el tamaño de la imagen base
        const canvas = createCanvas(baseImage.width, baseImage.height);
        const ctx = canvas.getContext('2d');

        // Dibujar la imagen base
        ctx.drawImage(baseImage, 0, 0);

        // Cargar fuente personalizada
        const fontPath = path.join(process.cwd(), 'assets', 'fonts', 'Comfortaa-Medium.ttf');
        const fontLoaded = loadFont(fontPath, 'Comfortaa');
        const font = fontLoaded ? '60px Comfortaa' : '60px Arial';

        // Escribir texto de bienvenida
        writeWelcomeText(ctx, nombreUsuario, font, baseImage.height);

        // Crear avatar circular con borde
        const avatarCanvas = makeRoundAvatarWithBorder(avatarImage);

        // Calcular posición del avatar (centrado verticalmente, a la izquierda)
        const avatarPositionY = (baseImage.height - avatarCanvas.height) / 2;

        // Dibujar el avatar en la imagen base
        ctx.drawImage(avatarCanvas, 25, avatarPositionY);

        // Retornar el buffer de la imagen final
        return canvas.toBuffer('image/png');
    } catch (error) {
        console.error(`Error al generar la imagen de bienvenida: ${error}`);
        return null;
    }
}

// MARK: Guardar imagen de bienvenida
/**
 * Función auxiliar para guardar la imagen en un archivo (útil para testing)
 * @param {string} nombreUsuario - Nombre del usuario
 * @param {string} avatarUrl - URL del avatar del usuario
 * @param {string} baseImagePath - Ruta de la imagen base
 * @param {string} outputPath - Ruta donde guardar la imagen generada
 * @returns {Promise<boolean>} - True si se guardó correctamente, false si hubo error
 */
export async function saveWelcomeImage(nombreUsuario, avatarUrl, baseImagePath, outputPath) {
    try {
        const imageBuffer = await generateWelcomeImage(nombreUsuario, avatarUrl, baseImagePath);
        if (!imageBuffer) {
            return false;
        }

        fs.writeFileSync(outputPath, imageBuffer);
        console.log(`Imagen de bienvenida guardada en: ${outputPath}`);
        return true;
    } catch (error) {
        console.error(`Error al guardar la imagen: ${error}`);
        return false;
    }
}

// MARK: Exportar funciones
// Exportar funciones individuales
export {
    cargarImagenBase,
    obtenerAvatar,
    makeRoundAvatarWithBorder,
    loadFont,
    writeWelcomeText
};
